### PR TITLE
Add proofs to Data.List.Properties

### PR DESCRIPTION
**Pull request by mildsunrise**
Hi! This is my first PR, apologies if I'm missing something.

This adds a series of proofs to `Data.List.Properties` that I consider general enough to be in the stdlib, and in most cases already exist for other list operations. This proves, in order:

 - interaction with `map`, for list operations where it makes sense and was missing
 - congruence for higher order operations, where missing (except for those acting on `Decidable`s, which I am still working on, and `unsnoc` which is challenging because of the indexed datatype)
 - for `alignWith` / `align`, `zipWith` / `zip` and `unzipWith` / `unzip`:
   - effect of swapping function arguments / inputs
   - equivalence of swapping function arguments if function is commutative
 - interaction of `take` with itself and with `drop`
 - `unzipWith` + `zipWith` being inverses (with particular cases for the non-With versions)
 - interaction with `++`, for operations where it makes sense and was missing

It also fixes a `module` that was defining a parameter that was not used in the proofs.
I've tried to respect conventions of that file where possible, please tell me if there's anything that could be improved :)

### CODE REVIEW

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:1412]: Hmm so you've made all the `f`s here and below explicit, but the existing proof `head-map` has an implicit `f`. Can `f` in general be inferable in these situations?

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:358]: In general with these `trans` proofs we usually prefer to use equational reasoning as it makes the proofs a little bit easier to read and follow?

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:382]: Eugh, I'm torn on this one whether this is a breaking change or not. As long as the user is not passing implicits explicitly it should be fine. Anyone else has have any feelings on it? 

If we decide it is fine, then we should definitely add a CHANGELOG entry for it.

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:130]: Why do we need the ih in a with? Why can't we just write `cong (y ∷_) (mapMaybe-cong f≗g xs)`? Similarly for proofs below?

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:323]: Nice spot, thank you!

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:372]: Let's not start sticking lists in module parameters? `∀ xs ys →` is way shorter both in the line and in terms of lines.

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:433]: Should this be generalised to `(∀ x y → f x y ≡ g y x)`?

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:449]: Again let's not put lists in module parameters.

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:674]: Having a proof of `d ≡ e` isn't very consistent with the other cong proofs above right? They omit propositional equalities, and only look at functional equalities.

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Hi @mildsunrise, thanks for the great first PR! Looks very good, but I have a few minor comments which I've added below.

**Comment by jamesmckinna** [comment]
Meta-comment: 
Should we fix `mapMaybe` first?
Cf. https://github.com/agda/agda/issues/5833 and links there to @omelkonian 's implementation?

**Comment by MatthewDaggitt** [comment]
What would "fixing" it involve?

**Comment by jamesmckinna** [comment]
> What would "fixing" it involve?

As in Orestis' prelude, write `catMaybe` *before* `mapMaybe` as the composition of `catMaybe` and `map`...

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:130]: I was mirroring the style of the existing proofs. I figured they put ih in a with to avoid repetiton in both branches, so I did the same here. I can change that if you prefer

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:382]: the diff is a little bit messy here, what happened is that [`zipWith-map` was added](https://github.com/agda/agda-stdlib/pull/2355/commits/76940a93cf9b1c36098f45988ca710a2731e1d53) (i put it at the start for consistency) and later [`zipWith-comm` was moved to the bottom](https://github.com/agda/agda-stdlib/pull/2355/commits/bb9a8a2c22266315175b2f61a36013a631e1f7e2) since it can be implemented more easily in terms of `zipWith-flip` and `zipWith-cong`, but there should be no breaking change

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:433]: that would be a breaking change (again, this was an existing proof that was moved to the bottom) but it can be done if you prefer

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:674]: yeah, that was my thought as well, but I preferred to mirror the structure of the existing `foldr-cong` proof that does accept `d ≡ e`. can be changed if you prefer

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:372]: ah, i see, i will do that. however keep in mind I cannot write `∀ xs ys →` as that would leave unbound list types. my reasoning was that the module should introduce what's necessary to bind list types A / B, but it is also reasoning that we not put lists in module parameters

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:1412]: right, I tried to figure out what to do from the existing `-map` proofs but they seem to be quite inconsistent, we sometimes use implicit and sometimes explicit. in most of them we do explicit, so that's what I went with for the new `-map` proofs.

as for whether it is in general inferable, it is hard for me to say. I usually prefer the resulting proof to be completely bound by the explicit parameters, but I don't have that much experience in Agda so I'll leave that call to you :)

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:358]: very reasonable, I'll do that here and in `unzip-flip` and `align-flip`

Comment (MatthewDaggitt) [reply src/Data/List/Properties.agda:358]: Yup there's various places in this file with a similar pattern.

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [reply src/Data/List/Properties.agda:382]: Ah sorry, my misunderstanding!

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [reply src/Data/List/Properties.agda:433]: Ah in which case, no it shouldn't be! Sorry thought it was a new proof!

**Comment by mildsunrise** [comment]
@MatthewDaggitt thanks for the thorough review! I've responded to your points and will act on them soon-ish, probably today

@jamesmckinna is it okay if I add the proposed fix to this PR, or would you rather keeping it in a separate PR, merging that one, and then rebasing this one on top?

**Comment by MatthewDaggitt** [comment]
>  As in Orestis' prelude, write catMaybe before mapMaybe as the composition of catMaybe and map... which avoids the awkward with construction...

Is this is a backwards compatible change with respect to reduction behaviour? We shouldn't do it now, if it's not.

**Comment by jamesmckinna** [comment]
I'll put my hobby-horse back in the stable... @MatthewDaggitt 's comment suggests we exercise caution. I'll investigate separately, and then take the knock-on hit from your additions downstream @mildsunrise so no action required on your part from me!

**Comment by mildsunrise** [comment]
to my understanding it should have equivalent reduction behavior, but yes, it's good to be cautious.
okay then, I'll fix the 2 remaining issues

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda]: done

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda]: done

**Comment by mildsunrise** [comment]
all done. let me know if you need anything else on my part, such as squashing the fixup commits into the original commits

**Comment by mildsunrise** [comment]
I've added a couple more properties, `mapMaybe-++` and `unzipWith-++` since their proof was similar enough to the existing proofs. If it feels like too much, I can leave them out of this PR

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:694]: see also [my comment](https://github.com/agda/agda-stdlib/issues/2358#issuecomment-2051480712) on #2358

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:314]: Can these be analysed using `foldr-++` etc.?

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:694]: ... and for having introduced a lemma whose name/type should be (downstream) deprecated in favour of the 'correct' version... oh dear!

File: CHANGELOG.md
Comment (mildsunrise) [reply CHANGELOG.md:314]: ooooh, good point, let me see

Comment (mildsunrise) [reply CHANGELOG.md:314]: oh, I see. do you have context about the "because pattern-matching definitions are generative" bit? i would like to learn more about why defining operations in Data.List.Base in terms of folds is a bad idea

Comment (JacquesCarette) [reply CHANGELOG.md:314]: What? Defining things in terms of folds should be a *good* idea!

Comment (jamesmckinna) [reply CHANGELOG.md:314]: Apologies for not being clear... but for details about generativity, you probably need to look at [the documentation](https://agda.readthedocs.io/en/latest/language/lambda-abstraction.html), which, somewhat frustratingly, doesn't make clear the relationship with [ordinary function definitions](https://agda.readthedocs.io/en/latest/language/function-definitions.html)...

That said, I have previously suggested refactoring the library to express (almost) all the recursions in `Data.List.Base` in terms of `foldr`/`foldl`, but IIRC, that was... discouraged. 

The following simple test suggests that, indeed, one should proceed with caution:
```agda
rev : List A → List A
rev = foldl (flip _∷_) []

rev′ : List A → List A
rev′ xs = go xs [] -- inline the definition using reverseAcc
  where
  go : List A → List A → List A
  go [] ys = ys
  go (x ∷ xs) ys = go xs (x ∷ ys)

test : rev ≡ rev′
test = refl
{- fails with:
foldl (λ y x → x ∷ y) [] x != Data.List.Base.go x x [] of type
List _A_1900
when checking that the expression refl has type rev ≡ rev′
-}
```

Comment (jamesmckinna) [reply CHANGELOG.md:314]: So I'm not saying it's a 'bad' idea, only that one changes the *intensional* semantics, even if not the *extensional* such. Sigh. 
For better or worse, the library privileges definitions made in direct pattern-matching style rather than by specialising higher-order operations...

Comment (jamesmckinna) [reply CHANGELOG.md:314]: See also #2267 / #2269 etc. which *do* change the intensional semantics (wrt strictness), but arguably towards 'better' definitions...

Comment (jamesmckinna) [reply CHANGELOG.md:314]: Last word: if #2355 get merged, we can fix up `mapMaybe-++` etc. downstream in #2361 assuming that gets approval...

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:130]: See also #2123 , second checkbox, on uses of 'irrefutable' `with` (when a `let` mucks up the flow of pattern-matching; LHS `let` will be a boon here)

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:132]: Having said all of that, my understand of @JacquesCarette 's point in #2359 is that such proof patterns might be (better?) finessed using the *dependent* function `Maybe.maybe`, in which case the `ih` terms might end up being able to be `let`-bound after all...?

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:162]: Ditto.

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:170]: ...

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:178]: ...

**Review by jamesmckinna — APPROVED** [review]
Happy with this as-is, but if the suggested use of `Maybe.maybe` permits the uses of `with` to be removed, then (probably?) so much the better! UPDATED: or not... so please ignore!

File: src/Data/List/Properties.agda
Comment (Taneb) [reply src/Data/List/Properties.agda:132]: The thing with `with`, as I understand it, is that proving things _about_ functions that use `with` can be inconvenient. We only very very rarely prove things about this kind of property. I think `with` is fine throughout this PR

Comment (JacquesCarette) [reply src/Data/List/Properties.agda:132]: My main 'beef' with `with` is when it is used to define functions, as that forces your hand downstream on how your proofs are forced to be done. And taking them out of proofs doesn't give as large a reward.

Having said that, I'm also interested in having proofs be less 'low level' (i.e. forcing explicit evaluation by following the definition patterns of the original function) and having more equational proofs. This would allow a smaller 'core API' that is representation dependent, and a larger set of properties which are representation *independent*.

So it's not so much 'simplification' of a proof as moving towards more intrinsic proofs.

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:694]: @MatthewDaggitt should i remove the `d ≡ e` or leave it in place to mirror `foldr-cong`? happy to do what you all prefer here

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda:132]: note that the use of `with` here mirrors the style of the existing `mapMaybe` proofs. if we want to move away from it, I'd prefer to leave it to a future PR that also changes the definition in Base

**Comment by mildsunrise** [comment]
(I'll be a bit busy these days, but I'd rather not block other PRs with this one, so please ping me if action is needed on my part)

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:132]: Absolutely! See #2361 although in the end I did *not* change the proof of `mapMaybe-concatMap`... but maybe I should!?

**Comment by jamesmckinna** [comment]
@MatthewDaggitt can you re-review and/or merge now, please? I think that this one is ready now ;-)

**Comment by gallais** [comment]
I think we just need the `d \== e` assumption to be removed and then we're good to merge.

File: src/Data/List/Properties.agda
Comment (MatthewDaggitt) [reply src/Data/List/Properties.agda:700]: Please could remove it? I think in the long run we want to try and remove it from the existing proofs as well.

**Comment by MatthewDaggitt** [comment]
Yup once removed, will merge :+1:

File: src/Data/List/Properties.agda
Comment (mildsunrise) [reply src/Data/List/Properties.agda]: done. `d ≡ e` was replaced by an (also explicit) `y` argument

**Comment by mildsunrise** [comment]
I see that #2361 was merged before this PR. given that `catMaybes` is no longer defined in terms of `mapMaybe`, it becomes necessary to prove properties about `catMaybes` directly. And once this is done, the properties of `mapMaybe` can be simplified to use those new proofs.

#2361 already did that with `length-mapMaybe` (added `length-catMaybes`). I've went ahead and done the same with the other three mapMaybe properties.

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**
